limitador/src/storage/in_memory.rs

```diff
@@ -35,36 +102,88 @@ impl CounterStorage for InMemoryStorage {
-                    ))
+        let mut limits_by_namespace = self.limits_for_namespace.write().unwrap();
+        let mut first_limited = None;
+
+        let mut process_counter =
+            |counter: &mut Counter, value: i64, delta: i64| -> Option<Authorization> {
+                if load_counters {
+                    let remaining = counter.max_value() - (value + delta);
+                    counter.set_remaining(remaining);
+                    if first_limited.is_none() && remaining < 0 {
+                        first_limited = Some(Authorization::Limited(
+                            counter.limit().name().map(|n| n.to_owned()),
+                        ));
+                    }
                 }
-            }
-            if let Some(l) = first_limited {
-                return Ok(l);
-            }
-        } else {
-            for counter in counters.iter() {
-                if !Self::counter_is_within_limits(counter, stored_counters.get(counter), delta) {
-                    return Ok(Authorization::Limited(
+                if !Self::counter_is_within_limits(counter, Some(&value), delta) {
+                    return Some(Authorization::Limited(
                         counter.limit().name().map(|n| n.to_owned()),
                     ));
                 }
+                None
+            };
+
+        for counter in counters.iter_mut() {
+            if counter.max_value() < delta {
+                if let Some(limited) = process_counter(counter, 0, delta) {
+                    if !load_counters {
+                        return Ok(limited);
+                    }
+                }
+                continue;
+            }
+            if let Some(limits) = limits_by_namespace.get(counter.limit().namespace()) {
+                if let Some(counters) = limits.get(counter.limit()) {
+                    if let Some(expiring_value) = counters.get(&counter.into()) {
+                        let value = expiring_value.value();
+                        if let Some(Authorization::Limited(counter_limited)) =
+                            process_counter(counter, value, delta)
+                        {
+                            if !load_counters {
+                                return Ok(Authorization::Limited(counter_limited));
+                            }
+                        }
+                    } else if let Some(limited) = process_counter(counter, 0, delta) {
+                        if !load_counters {
+                            return Ok(limited);
+                        }
+                    }
+                } else if let Some(limited) = process_counter(counter, 0, delta) {
+                    if !load_counters {
+                        return Ok(limited);
+                    }
+                }
+            } else if let Some(limited) = process_counter(counter, 0, delta) {
+                if !load_counters {
+                    return Ok(limited);
```

Comment:
I think we could easily avoid this duplication here, and only do this once, maybe after this entire logic…